apollo-federation/src/sources/connect/models/validation/source_name.rs

```diff
@@ -0,0 +1,67 @@
+use apollo_compiler::ast::Argument;
+use apollo_compiler::ast::Value;
+use apollo_compiler::Name;
+use apollo_compiler::Node;
+use apollo_compiler::SourceMap;
+
+use super::source_name_value_coordinate;
+use super::Code;
+use super::Location;
+use super::Message;
+use super::SourceName;
+use crate::sources::connect::spec::schema::CONNECT_SOURCE_ARGUMENT_NAME;
+
+pub(super) fn validate_source_name_arg(
+    field_name: &Name,
+    object_name: &Name,
+    source_name: &Node<Argument>,
+    source_map: &SourceMap,
+    source_names: &[SourceName],
+    source_directive_name: &Name,
+    connect_directive_name: &Name,
+) -> Vec<Message> {
+    let mut messages = vec![];
+
+    if source_names.iter().all(|name| name != &source_name.value) {
+        // TODO: Pick a suggestion that's not just the first defined source
+        let qualified_directive = connect_directive_name_coordinate(
+            connect_directive_name,
+            &source_name.value,
+            object_name,
+            field_name,
+        );
+        if let Some(first_source_name) = source_names.first() {
+            messages.push(Message {
+                    code: Code::SourceNameMismatch,
+                    message: format!(
+                        "{qualified_directive} does not match any defined sources. Did you mean {first_source_name}?",
+                    ),
+                    locations: Location::from_node(source_name.location(), source_map)
+                        .into_iter()
+                        .collect(),
+                });
+        } else {
+            messages.push(Message {
+                    code: Code::NoSourcesDefined,
+                    message: format!(
+                        "{qualified_directive} specifies a source, but none are defined. Try adding {coordinate} to the schema.",
+                        coordinate = source_name_value_coordinate(source_directive_name, &source_name.value),
+                    ),
+                    locations: Location::from_node(source_name.location(), source_map)
+                        .into_iter()
+                        .collect(),
+                });
+        }
+    }
+
+    messages
+}
+
+fn connect_directive_name_coordinate(
```

Comment:
We should keep all the coordinates together, I think, for easier discovery. Maybe in their own `schema_coordinates` module. Still tinkering with the idea of making them into structs instead of format strings, which I think could be a bit more efficient 🤔.

Reply:
The way this is structured now is that there is a top level `validate(Schema)` method, that knows how to validate everything. Even if some bits of code are moved into sub-modules, the top level method still calls into those modules directly in a tightly coupled way, passing in the exact arguments they need.

Instead, the way I've generally seen this kind of validation done is to decouple the validation engine from the individual validators. The top level `validate` method just runs a validation engine, that iterates over the Schema nodes and calls out to any validators that have registered an interest in the current node.

So a validator might indicate that it's interested in any field with a `@connect` directive (or maybe more specifically, a `@connect` directive with a `headers` attribute). When the validation engine sees such a node, it calls the validator and passes in that field. The validator implementation then does whatever it needs to do.

Adding in a new validator is then very decoupled - no changes are needed to the engine itself, you just add a new validator and usually a line of code to register the new validator.

This might all be overkill for now, since we're just looking at a couple of directives, but as things grow this kind of structure can be beneficial.

Reply:
That sort of abstraction could definitely be useful if it makes it easier to read the code. It kind of gets at my question around overall structure above. If we organize by individual validation rule, then we probably want some pattern that decouples from the schema structure. If we organize by schema structure, I think that level of decoupling wouldn't provide the same value.

We also only have the one consumer, so whatever organization/abstraction we go for here only needs to be optimized for readability, not flexibility.